frontend/src/metabase-lib/lib/Dimension.js

```diff
@@ -604,6 +604,17 @@ export class FieldDimension extends Dimension {
       }
     }
 
+    let fieldMetadata = {};
+
+    // If a field dimension isn't associated with a query,
+    // there is a chance it's stored in fields metadata
+    if (this.isStringFieldName()) {
+      // As field literal's IDs are not numeric like in field references,
+      // in metadata.fields their IDs are stringified MBQL objects
+      const metadataFieldLiteralId = String(this.mbql());
+      fieldMetadata = this._metadata?.field(metadataFieldLiteralId) || {};
+    }
```

Comment:
The `String(this.mbql())` part is a bit weird, but this is how field literals are stored in metadata's field object. See an example:

<details>
<summary>Screenshot</summary>

<img width="621" alt="field ids" src="https://user-images.githubusercontent.com/17258145/140975235-aba3e4ae-ce79-4bc8-973e-3b98fd5720bd.png">

</details>

When the `fields` object is created, we use the field's `id` as a key. For field literals, the `id` contains the whole MBQL object, so it gets coerced to a string and then used as a key. A more clean solution would also fix the way we store field IDs here, but I'm keeping the fix as minimal as possible to drop the risks

Reply:
I don't think we should use the stringified mbql keys found in `metadata.fields` because they aren't unique identifiers. TBH, we should probably filter them out of this map entirely to prevent usage since I think their inclusion with keys like that is a bug. I tried remapping them to `${tableId}:${name}` a while back but had some problems so abandoned that (for now at least).

Fields like this, fortunately, ought to show up on a `query` in the `result_metadata` property. I'm not sure why this code is skipping the giant `if` block right above this. My guess is that this `dimension` wasn't passed a `query` in the third arg of `Dimension.parseMBQL(...)` when it was created -- we really ought to always be calling `Dimension.parseMBQL(mbql, metadata, query)` but there seem to be cases where we are just using the first one or two args. Another possibility is that the logic for accessing `result_metadata` isn't quite right.